include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -0,0 +1,138 @@
+
+#ifndef __NBL_ASSET_I_SHADER_COMPILER_H_INCLUDED__
+#define __NBL_ASSET_I_SHADER_COMPILER_H_INCLUDED__
+
+#include "nbl/core/declarations.h"
+#include "nbl/system/declarations.h"
+
+#include "nbl/system/IFile.h"
+#include "nbl/system/ISystem.h"
+
+#include "nbl/asset/ICPUSpecializedShader.h"
+#include "nbl/asset/utils/IIncludeHandler.h"
+
+#include "nbl/asset/utils/ISPIRVOptimizer.h"
+
+namespace nbl::asset
+{
+
+class NBL_API IShaderCompiler : public core::IReferenceCounted
+{
+	public:
+
+		IShaderCompiler(system::ISystem* _s);
+
+		/**
+		Resolves ALL #include directives regardless of any other preprocessor directive.
+		This is done in order to support `#include` AND simultaneulsy be able to store (serialize) such ICPUShader (mostly GLSL source) into ONE file which, upon loading, will compile on every hardware/driver predicted by shader's author.
+
+		Internally function "disables" all preprocessor directives (so that they're not processed by preprocessor) except `#include` (and also `#version` and `#pragma shader_stage`).
+		Note that among the directives there may be include guards. Because of that, _maxSelfInclusionCnt parameter is provided.
+
+		@param _maxSelfInclusionCnt Max self-inclusion count of possible file being #include'd. If no self-inclusions are allowed, should be set to 0.
+
+		@param _originFilepath Path to not necesarilly existing file whose directory will be base for relative (""-type) top-level #include's resolution.
+			If _originFilepath is non-path-like string (e.g. "whatever" - no slashes), the base directory is assumed to be "." (working directory of your executable). It's important for it to be unique.
+
+		@returns Shader containing logically same GLSL code as input but with #include directives resolved.
+		*/
+		core::smart_refctd_ptr<ICPUShader> resolveIncludeDirectives(
+			std::string&& _code,
+			IShader::E_SHADER_STAGE _stage,
+			const char* _originFilepath,
+			uint32_t _maxSelfInclusionCnt = 4u,
+			system::logger_opt_ptr logger = nullptr) const;
+
+		core::smart_refctd_ptr<ICPUShader> resolveIncludeDirectives(
+			system::IFile* _sourcefile,
+			IShader::E_SHADER_STAGE _stage,
+			const char* _originFilepath,
+			uint32_t _maxSelfInclusionCnt = 4u,
+			system::logger_opt_ptr logger = nullptr) const;
+		
+		/*
+			Creates a formatted copy of the original
+
+			@param original An original glsl shader (must contain glsl and not be a dummy object of be a nullptr).
+			@param fmt A string with c-like format, which will be filled with data from ...args
+			@param ...args Data to fill fmt with
+			@returns shader containing fmt filled with ...args, placed before the original code.
+
+			If original == nullptr, the output buffer will only contain the data from fmt. If original code contains #version specifier,
+			then the filled fmt will be placed onto the next line after #version in the output buffer. If not, fmt will be placed into the
+			beginning of the output buffer.
```

Comment:
to make this work both for HLSL and GLSL, we need to expose the position where we'll paste the formatted string and give GLSL compiler a function to find the first line after the `#version` directive and an overload of this static function which misses the position parameter

Reply:
I'll leave this for the next PR for whoever works on integrating DXC and compiling HLSL
Note that this PR is only to have IShaderCompiler cleaned up and nice + IncludeHandler related files
and only the interface for HLSLCompiler not the implementation. then I'll be done. (I'll update the PR notes)